src/pass/type_infer.cc

```diff
@@ -132,44 +137,52 @@ class TypeInferencer : public ExprMutator {
       UpdateFuncParamVarMap(fn, call->args);
     }
 
-    Expr op = VisitExpr(call->op);
-    Call ret = Call(op, args, call->attrs, call->type_args);
-    if (const FunctionNode* fn = ret->op.as<FunctionNode>()) {
-      ret->checked_type_ = InferClosure(ret);
-    } else if (const GlobalVarNode* gvn = ret->op.as<GlobalVarNode>()) {
-      ret->op->checked_type_ =
-          Unify(gvn->checked_type(), mod_->Lookup(GetRef<GlobalVar>(gvn))->checked_type());
-      ret->checked_type_ = InferClosure(ret);
-    } else if (const OpNode* opn = ret->op.as<OpNode>()) {
-      ret->checked_type_ = InferPrimitive(ret, opn);
-    } else if (ret->op.as<VarNode>() || ret->op.as<LetNode>()) {
-      // handle recursive func call when op is a var node
-      if (op->checked_type()->IsInstance<IncompleteTypeNode>()) {
-        ret->checked_type_ = IncompleteType(kType);
-      } else {
-        // The var node can be a result of the output type of a func call. A var node
-        // here is valid if it points to a function. Check that the type is a FuncType
-        // and the args of the Call match the type of the FuncType. If yes, return the
-        // FuncType's ret_type.
-        if (const auto* var_node = ret->op.as<VarNode>()) {
-          VisitPrimitiveClosureFromCallerArgs(var_node, call->args);
-        }
-        const FuncTypeNode* fty_node = ret->op->checked_type_.as<FuncTypeNode>();
-        CHECK(fty_node);
-        for (size_t i = 0; i < fty_node->arg_types.size(); i++) {
-          ret->args[i]->checked_type_ = Unify(fty_node->arg_types[i], ret->args[i]->checked_type());
+    Call ret = Call(call->op, args, call->attrs, call->type_args);
+    if (const FunctionNode* fn = call->op.as<FunctionNode>()) {
+      auto ret_type = InferClosure(ret, GetRef<Function>(fn));
+      ret = Call(VisitExpr(call->op), args, call->attrs, call->type_args);
+      ret->checked_type_ = ret_type;
+    } else if (const GlobalVarNode* gvn = call->op.as<GlobalVarNode>()) {
+      auto fn = Downcast<Function>(mod_->Lookup(GetRef<GlobalVar>(gvn)));
+      auto ret_type = InferClosure(ret, fn);
+      ret = Call(VisitExpr(call->op), args, call->attrs, call->type_args);
```

Comment:
what does `VisitExpr(call->op)` do?

Reply:
In the case that the closure has been updated with the new param type by `InferClosure`, visit the closure here will give you the updated one.

Reply:
`call->op` here is a globalvar, IIUC, the global function will not be updated (and should not?) and we have no such case now?

Reply:
Hmm that's true. In the case of global var this is unnecessary.